File: compat/suites/go-sdk/internal/groups/kms.go
package groups

import (
	"context"
	"fmt"

	"github.com/Neaox/overcast-compat-go-sdk/internal/clients"
	"github.com/Neaox/overcast-compat-go-sdk/internal/harness"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/kms"
	"github.com/aws/aws-sdk-go-v2/service/kms/types"
)

func KMS(c *clients.Clients) ServiceGroup {
	g := &kmsGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateKey":                       g.CreateKey,
			"DescribeKey":                     g.DescribeKey,
			"ListKeys":                        g.ListKeys,
			"EnableKey":                       g.EnableKey,
			"DisableKey":                      g.DisableKey,
			"Encrypt":                         g.Encrypt,
			"Decrypt":                         g.Decrypt,
			"GenerateDataKey":                 g.GenerateDataKey,
			"GenerateDataKeyWithoutPlaintext": g.GenerateDataKeyWithoutPlaintext,
			"TagKMSResource":                  g.TagKMSResource,
			"UntagKMSResource":                g.UntagKMSResource,
			"ListKMSResourceTags":             g.ListKMSResourceTags,
			"ScheduleKeyDeletion":             g.ScheduleKeyDeletion,
			"CancelKeyDeletion":               g.CancelKeyDeletion,
			"Sign":                            g.Sign,
			"Verify":                          g.Verify,
			"CreateKmsAlias":                  g.CreateKmsAlias,
			"ListKmsAliases":                  g.ListKmsAliases,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"kms-keys":       g.setupKeys,
			"kms-crypto":     g.setupCrypto,
			"kms-asymmetric": g.setupAsymmetric,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"kms-keys":       g.teardownKeys,
			"kms-crypto":     g.teardownCrypto,
			"kms-asymmetric": g.teardownAsymmetric,
		},
	}
}

type kmsGroup struct{ c *clients.Clients }

func (g *kmsGroup) cl() *kms.Client { return g.c.KMS() }

// ── kms-keys ──────────────────────────────────────────────────────────────────

func (g *kmsGroup) setupKeys(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().CreateKey(ctx, &kms.CreateKeyInput{
		Description: aws.String(fmt.Sprintf("oc-key-%s", t.RunID)),
	})
	if err != nil {
		return err
	}
	t.Set("kms_key_id", aws.ToString(resp.KeyMetadata.KeyId))
	return nil
}

func (g *kmsGroup) teardownKeys(ctx context.Context, t *harness.TestContext) error {
	if keyID := t.GetString("kms_key_id"); keyID != "" {
		// Delete alias before scheduling key deletion — alias outlives the key otherwise.
		g.cl().DeleteAlias(ctx, &kms.DeleteAliasInput{ //nolint:errcheck
			AliasName: aws.String(fmt.Sprintf("alias/compat-%s", t.RunID)),
		})
		g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
			KeyId:               aws.String(keyID),
			PendingWindowInDays: aws.Int32(7),
		}) //nolint:errcheck
	}
	if keyID := t.GetString("kms_pending_key_id"); keyID != "" {
		g.cl().CancelKeyDeletion(ctx, &kms.CancelKeyDeletionInput{KeyId: aws.String(keyID)}) //nolint:errcheck
		g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
			KeyId: aws.String(keyID), PendingWindowInDays: aws.Int32(7),
		}) //nolint:errcheck
	}
	return nil
}

func (g *kmsGroup) CreateKey(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().CreateKey(ctx, &kms.CreateKeyInput{
		Description: aws.String("oc-test-key"),
	})
	if err == nil {
		g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
			KeyId: resp.KeyMetadata.KeyId, PendingWindowInDays: aws.Int32(7),
		}) //nolint:errcheck
	}
	return err
}

func (g *kmsGroup) DescribeKey(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().DescribeKey(ctx, &kms.DescribeKeyInput{
		KeyId: aws.String(t.GetString("kms_key_id")),
	})
	if err != nil {
		return err
	}
	if resp.KeyMetadata == nil {
		return fmt.Errorf("DescribeKey: nil metadata")
	}
	return nil
}

func (g *kmsGroup) ListKeys(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().ListKeys(ctx, &kms.ListKeysInput{})
	if err != nil {
		return err
	}
	if len(resp.Keys) == 0 {
		return fmt.Errorf("ListKeys: expected ≥1 key")
	}
	return nil
}

func (g *kmsGroup) EnableKey(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().EnableKey(ctx, &kms.EnableKeyInput{KeyId: aws.String(t.GetString("kms_key_id"))})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeKey(ctx, &kms.DescribeKeyInput{KeyId: aws.String(t.GetString("kms_key_id"))})
	if err != nil {
		return fmt.Errorf("EnableKey: DescribeKey verify failed: %w", err)
	}
	if !resp.KeyMetadata.Enabled {
		return fmt.Errorf("EnableKey: key not enabled after EnableKey")
	}
	return nil
}

func (g *kmsGroup) DisableKey(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().DisableKey(ctx, &kms.DisableKeyInput{KeyId: aws.String(t.GetString("kms_key_id"))})
	if err != nil {
		return err
	}
	resp, err := g.cl().DescribeKey(ctx, &kms.DescribeKeyInput{KeyId: aws.String(t.GetString("kms_key_id"))})
	if err != nil {
		return fmt.Errorf("DisableKey: DescribeKey verify failed: %w", err)
	}
	if resp.KeyMetadata.Enabled {
		return fmt.Errorf("DisableKey: key still enabled after DisableKey")
	}
	// re-enable for subsequent tests
	g.cl().EnableKey(ctx, &kms.EnableKeyInput{KeyId: aws.String(t.GetString("kms_key_id"))}) //nolint:errcheck
	return nil
}

func (g *kmsGroup) TagKMSResource(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().TagResource(ctx, &kms.TagResourceInput{
		KeyId: aws.String(t.GetString("kms_key_id")),
		Tags: []types.Tag{
			{TagKey: aws.String("env"), TagValue: aws.String("test")},
			{TagKey: aws.String("run"), TagValue: aws.String(t.RunID)},
		},
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ListResourceTags(ctx, &kms.ListResourceTagsInput{
		KeyId: aws.String(t.GetString("kms_key_id")),
	})
	if err != nil {
		return fmt.Errorf("TagKMSResource: ListResourceTags verify failed: %w", err)
	}
	for _, tag := range resp.Tags {
		if aws.ToString(tag.TagKey) == "env" && aws.ToString(tag.TagValue) == "test" {
			return nil
		}
	}
	return fmt.Errorf("TagKMSResource: env=test tag not found")
}

func (g *kmsGroup) UntagKMSResource(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().UntagResource(ctx, &kms.UntagResourceInput{
		KeyId:   aws.String(t.GetString("kms_key_id")),
		TagKeys: []string{"run"},
	})
	return err
}

func (g *kmsGroup) ListKMSResourceTags(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().ListResourceTags(ctx, &kms.ListResourceTagsInput{
		KeyId: aws.String(t.GetString("kms_key_id")),
	})
	if err != nil {
		return err
	}
	if len(resp.Tags) == 0 {
		return fmt.Errorf("ListKMSResourceTags: expected ≥1 tag")
	}
	return nil
}

func (g *kmsGroup) ScheduleKeyDeletion(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().CreateKey(ctx, &kms.CreateKeyInput{
		Description: aws.String("oc-sched-del"),
	})
	if err != nil {
		return err
	}
	keyID := aws.ToString(resp.KeyMetadata.KeyId)
	t.Set("kms_pending_key_id", keyID)
	_, err = g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
		KeyId:               aws.String(keyID),
		PendingWindowInDays: aws.Int32(7),
	})
	return err
}

func (g *kmsGroup) CancelKeyDeletion(ctx context.Context, t *harness.TestContext) error {
	keyID := t.GetString("kms_pending_key_id")
	if keyID == "" {
		return nil
	}
	_, err := g.cl().CancelKeyDeletion(ctx, &kms.CancelKeyDeletionInput{KeyId: aws.String(keyID)})
	return err
}

func (g *kmsGroup) CreateKmsAlias(ctx context.Context, t *harness.TestContext) error {
	aliasName := fmt.Sprintf("alias/compat-%s", t.RunID)
	_, err := g.cl().CreateAlias(ctx, &kms.CreateAliasInput{
		AliasName:   aws.String(aliasName),
		TargetKeyId: aws.String(t.GetString("kms_key_id")),
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ListAliases(ctx, &kms.ListAliasesInput{KeyId: aws.String(t.GetString("kms_key_id"))})
	if err != nil {
		return fmt.Errorf("CreateKmsAlias: ListAliases verify failed: %w", err)
	}
	for _, a := range resp.Aliases {
		if aws.ToString(a.AliasName) == aliasName {
			return nil
		}
	}
	return fmt.Errorf("CreateKmsAlias: alias %q not found", aliasName)
}

func (g *kmsGroup) ListKmsAliases(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().ListAliases(ctx, &kms.ListAliasesInput{
		KeyId: aws.String(t.GetString("kms_key_id")),
	})
	if err != nil {
		return err
	}
	aliasName := fmt.Sprintf("alias/compat-%s", t.RunID)
	for _, a := range resp.Aliases {
		if aws.ToString(a.AliasName) == aliasName {
			return nil
		}
	}
	return fmt.Errorf("ListAliases: alias %q not found", aliasName)
}

// ── kms-crypto ────────────────────────────────────────────────────────────────

func (g *kmsGroup) setupCrypto(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().CreateKey(ctx, &kms.CreateKeyInput{
		Description: aws.String(fmt.Sprintf("oc-crypto-%s", t.RunID)),
	})
	if err != nil {
		return err
	}
	t.Set("kms_crypto_key_id", aws.ToString(resp.KeyMetadata.KeyId))
	return nil
}

func (g *kmsGroup) teardownCrypto(ctx context.Context, t *harness.TestContext) error {
	if keyID := t.GetString("kms_crypto_key_id"); keyID != "" {
		g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
			KeyId: aws.String(keyID), PendingWindowInDays: aws.Int32(7),
		}) //nolint:errcheck
	}
	return nil
}

func (g *kmsGroup) Encrypt(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().Encrypt(ctx, &kms.EncryptInput{
		KeyId:     aws.String(t.GetString("kms_crypto_key_id")),
		Plaintext: []byte("hello world"),
	})
	if err != nil {
		return err
	}
	t.Set("kms_ciphertext", resp.CiphertextBlob)
	return nil
}

func (g *kmsGroup) Decrypt(ctx context.Context, t *harness.TestContext) error {
	ct, ok := t.Get("kms_ciphertext")
	if !ok {
		// encrypt first
		encResp, err := g.cl().Encrypt(ctx, &kms.EncryptInput{
			KeyId: aws.String(t.GetString("kms_crypto_key_id")), Plaintext: []byte("hello world"),
		})
		if err != nil {
			return err
		}
		ct = encResp.CiphertextBlob
	}
	resp, err := g.cl().Decrypt(ctx, &kms.DecryptInput{CiphertextBlob: ct.([]byte)})
	if err != nil {
		return err
	}
	if string(resp.Plaintext) != "hello world" {
		return fmt.Errorf("Decrypt: expected plaintext %q, got %q", "hello world", string(resp.Plaintext))
	}
	return nil
}

func (g *kmsGroup) GenerateDataKey(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GenerateDataKey(ctx, &kms.GenerateDataKeyInput{
		KeyId:   aws.String(t.GetString("kms_crypto_key_id")),
		KeySpec: types.DataKeySpecAes256,
	})
	if err != nil {
		return err
	}
	if len(resp.Plaintext) == 0 {
		return fmt.Errorf("GenerateDataKey: empty plaintext")
	}
	return nil
}

func (g *kmsGroup) GenerateDataKeyWithoutPlaintext(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GenerateDataKeyWithoutPlaintext(ctx, &kms.GenerateDataKeyWithoutPlaintextInput{
		KeyId:   aws.String(t.GetString("kms_crypto_key_id")),
		KeySpec: types.DataKeySpecAes256,
	})
	if err != nil {
		return err
	}
	if len(resp.CiphertextBlob) == 0 {
		return fmt.Errorf("GenerateDataKeyWithoutPlaintext: empty ciphertext")
	}
	return nil
}

func (g *kmsGroup) GenerateRandom(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().GenerateRandom(ctx, &kms.GenerateRandomInput{
		NumberOfBytes: aws.Int32(32),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	if len(resp.Plaintext) == 0 {
		return fmt.Errorf("GenerateRandom: empty bytes")
	}
	return nil
}

// ── kms-asymmetric ────────────────────────────────────────────────────────────

func (g *kmsGroup) setupAsymmetric(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().CreateKey(ctx, &kms.CreateKeyInput{
		Description: aws.String(fmt.Sprintf("oc-asym-%s", t.RunID)),
		KeySpec:     types.KeySpecRsa2048,
		KeyUsage:    types.KeyUsageTypeSignVerify,
	})
	if err != nil {
		return err
	}
	t.Set("kms_asym_key_id", aws.ToString(resp.KeyMetadata.KeyId))
	return nil
}

func (g *kmsGroup) teardownAsymmetric(ctx context.Context, t *harness.TestContext) error {
	if keyID := t.GetString("kms_asym_key_id"); keyID != "" {
		g.cl().ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
			KeyId: aws.String(keyID), PendingWindowInDays: aws.Int32(7),
		}) //nolint:errcheck
	}
	return nil
}

func (g *kmsGroup) Sign(ctx context.Context, t *harness.TestContext) error {
	keyID := t.GetString("kms_asym_key_id")
	resp, err := g.cl().Sign(ctx, &kms.SignInput{
		KeyId:            aws.String(keyID),
		Message:          []byte("test data"),
		MessageType:      types.MessageTypeRaw,
		SigningAlgorithm: types.SigningAlgorithmSpecRsassaPkcs1V15Sha256,
	})
	if err != nil {
		return err
	}
	if len(resp.Signature) == 0 {
		return fmt.Errorf("Sign: empty signature")
	}
	t.Set("kms_signature", resp.Signature)
	return nil
}

func (g *kmsGroup) Verify(ctx context.Context, t *harness.TestContext) error {
	keyID := t.GetString("kms_asym_key_id")
	sig, ok := t.Get("kms_signature")
	if !ok {
		// Sign must run first; sign inline
		signResp, err := g.cl().Sign(ctx, &kms.SignInput{
			KeyId:            aws.String(keyID),
			Message:          []byte("test data"),
			MessageType:      types.MessageTypeRaw,
			SigningAlgorithm: types.SigningAlgorithmSpecRsassaPkcs1V15Sha256,
		})
		if err != nil {
			return err
		}
		sig = signResp.Signature
	}
	resp, err := g.cl().Verify(ctx, &kms.VerifyInput{
		KeyId:            aws.String(keyID),
		Message:          []byte("test data"),
		MessageType:      types.MessageTypeRaw,
		Signature:        sig.([]byte),
		SigningAlgorithm: types.SigningAlgorithmSpecRsassaPkcs1V15Sha256,
	})
	if err != nil {
		return err
	}
	if !resp.SignatureValid {
		return fmt.Errorf("Verify: signature not valid")
	}
	return nil
}

File: compat/suites/python-sdk/groups/kms.py
"""
groups/kms.py — KMS compatibility test implementations for the Python suite.
"""

from __future__ import annotations
import base64
from lib.harness import TestContext
from lib.clients import make_clients


def _kms(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).kms


# ── kms-keys ──────────────────────────────────────────────────────────────────

def setup_kms_keys(ctx: TestContext) -> None:
    kms = _kms(ctx)
    resp = kms.create_key(Description=f"compat-test {ctx.run_id}")
    ctx["kms_key_id"] = resp["KeyMetadata"]["KeyId"]
    ctx["kms_key_arn"] = resp["KeyMetadata"]["Arn"]


def teardown_kms_keys(ctx: TestContext) -> None:
    key_id = ctx.get("kms_key_id")
    if key_id:
        kms = _kms(ctx)
        # Delete alias first — it outlives the key otherwise.
        alias = ctx.get("kms_alias")
        if alias:
            try:
                kms.delete_alias(AliasName=alias)
            except Exception:
                pass
        try:
            kms.schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
        except Exception:
            pass


def CreateKey(ctx: TestContext) -> None:
    kms = _kms(ctx)
    resp = kms.create_key(Description=f"compat temp {ctx.run_id}")
    key_id = resp["KeyMetadata"]["KeyId"]
    try:
        if not key_id:
            raise AssertionError("CreateKey: missing KeyId")
    finally:
        try:
            kms.schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
        except Exception:
            pass


def DescribeKey(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    resp = kms.describe_key(KeyId=key_id)
    if resp["KeyMetadata"]["KeyId"] != key_id:
        raise AssertionError(f"DescribeKey: wrong KeyId {resp['KeyMetadata']['KeyId']!r}")


def CreateAlias(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    alias = f"alias/{ctx.run_id}-compat"
    kms.create_alias(AliasName=alias, TargetKeyId=key_id)
    ctx["kms_alias"] = alias
    resp = kms.list_aliases()
    names = [a["AliasName"] for a in resp.get("Aliases", [])]
    assert alias in names, f"CreateAlias: alias {alias} not found in list"


def ListAliases(ctx: TestContext) -> None:
    kms = _kms(ctx)
    alias = ctx.get("kms_alias")
    resp = kms.list_aliases()
    names = [a["AliasName"] for a in resp.get("Aliases", [])]
    if alias and alias not in names:
        raise AssertionError(f"ListAliases: {alias!r} not found in aliases")


def ListKeys(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    resp = kms.list_keys()
    key_ids = [k["KeyId"] for k in resp.get("Keys", [])]
    if key_id not in key_ids:
        raise AssertionError(f"ListKeys: {key_id!r} not found in key list")


def DisableKey(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    kms.disable_key(KeyId=key_id)
    resp = kms.describe_key(KeyId=key_id)
    if resp["KeyMetadata"]["KeyState"] != "Disabled":
        raise AssertionError(f"DisableKey: expected Disabled, got {resp['KeyMetadata']['KeyState']!r}")


def EnableKey(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    kms.enable_key(KeyId=key_id)
    resp = kms.describe_key(KeyId=key_id)
    if resp["KeyMetadata"]["KeyState"] != "Enabled":
        raise AssertionError(f"EnableKey: expected Enabled, got {resp['KeyMetadata']['KeyState']!r}")


def ScheduleKeyDeletion(ctx: TestContext) -> None:
    kms = _kms(ctx)
    # Create a fresh key to schedule for deletion
    resp = kms.create_key(Description="to-delete")
    key_id = resp["KeyMetadata"]["KeyId"]
    kms.schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
    desc = kms.describe_key(KeyId=key_id)
    if desc["KeyMetadata"]["KeyState"] != "PendingDeletion":
        raise AssertionError(f"ScheduleKeyDeletion: expected PendingDeletion, got {desc['KeyMetadata']['KeyState']!r}")
    ctx["kms_pending_key_id"] = key_id


def CancelKeyDeletion(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx.get("kms_pending_key_id")
    if not key_id:
        raise AssertionError("CancelKeyDeletion: no pending key")
    kms.cancel_key_deletion(KeyId=key_id)
    desc = kms.describe_key(KeyId=key_id)
    state = desc["KeyMetadata"]["KeyState"]
    assert state == "Disabled", f"CancelKeyDeletion: expected Disabled, got {state}"
    try:
        kms.schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
    except Exception:
        pass


def TagKMSResource(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    kms.tag_resource(
        KeyId=key_id,
        Tags=[{"TagKey": "env", "TagValue": "test"}],
    )


def ListKMSResourceTags(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    resp = kms.list_resource_tags(KeyId=key_id)
    tags = resp.get("Tags", [])
    found = any(t["TagKey"] == "env" for t in tags)
    if not found:
        raise AssertionError("ListKMSResourceTags: tag 'env' not found")


def UntagKMSResource(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_key_id"]
    kms.untag_resource(KeyId=key_id, TagKeys=["env"])


# ── kms-crypto ────────────────────────────────────────────────────────────────

def setup_kms_crypto(ctx: TestContext) -> None:
    kms = _kms(ctx)
    resp = kms.create_key(Description=f"crypto-test {ctx.run_id}")
    ctx["kms_crypto_key_id"] = resp["KeyMetadata"]["KeyId"]


def teardown_kms_crypto(ctx: TestContext) -> None:
    key_id = ctx.get("kms_crypto_key_id")
    if key_id:
        try:
            _kms(ctx).schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
        except Exception:
            pass


def Encrypt(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_crypto_key_id"]
    resp = kms.encrypt(KeyId=key_id, Plaintext=b"hello kms")
    if not resp.get("CiphertextBlob"):
        raise AssertionError("Encrypt: missing CiphertextBlob")
    ctx["kms_ciphertext"] = resp["CiphertextBlob"]


def Decrypt(ctx: TestContext) -> None:
    kms = _kms(ctx)
    ciphertext = ctx.get("kms_ciphertext")
    if not ciphertext:
        raise AssertionError("Decrypt: no ciphertext available")
    resp = kms.decrypt(CiphertextBlob=ciphertext)
    plaintext = resp.get("Plaintext", b"")
    if plaintext != b"hello kms":
        raise AssertionError(f"Decrypt: expected b'hello kms', got {plaintext!r}")


def GenerateDataKey(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_crypto_key_id"]
    resp = kms.generate_data_key(KeyId=key_id, KeySpec="AES_256")
    if not resp.get("Plaintext"):
        raise AssertionError("GenerateDataKey: missing Plaintext")
    if not resp.get("CiphertextBlob"):
        raise AssertionError("GenerateDataKey: missing CiphertextBlob")


def GenerateDataKeyWithoutPlaintext(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_crypto_key_id"]
    resp = kms.generate_data_key_without_plaintext(KeyId=key_id, KeySpec="AES_256")
    if not resp.get("CiphertextBlob"):
        raise AssertionError("GenerateDataKeyWithoutPlaintext: missing CiphertextBlob")
    if resp.get("Plaintext"):
        raise AssertionError("GenerateDataKeyWithoutPlaintext: should not return Plaintext")


# ── kms-asymmetric ────────────────────────────────────────────────────────────

def setup_kms_asymmetric(ctx: TestContext) -> None:
    kms = _kms(ctx)
    resp = kms.create_key(
        KeySpec="RSA_2048",
        KeyUsage="SIGN_VERIFY",
        Description=f"asymmetric {ctx.run_id}",
    )
    ctx["kms_asym_key_id"] = resp["KeyMetadata"]["KeyId"]


def teardown_kms_asymmetric(ctx: TestContext) -> None:
    key_id = ctx.get("kms_asym_key_id")
    if key_id:
        try:
            _kms(ctx).schedule_key_deletion(KeyId=key_id, PendingWindowInDays=7)
        except Exception:
            pass


def Sign(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_asym_key_id"]
    resp = kms.sign(
        KeyId=key_id,
        Message=b"message to sign",
        MessageType="RAW",
        SigningAlgorithm="RSASSA_PKCS1_V1_5_SHA_256",
    )
    if not resp.get("Signature"):
        raise AssertionError("Sign: missing Signature")
    ctx["kms_signature"] = resp["Signature"]


def Verify(ctx: TestContext) -> None:
    kms = _kms(ctx)
    key_id = ctx["kms_asym_key_id"]
    sig = ctx.get("kms_signature")
    if not sig:
        raise AssertionError("Verify: no signature available")
    resp = kms.verify(
        KeyId=key_id,
        Message=b"message to sign",
        MessageType="RAW",
        Signature=sig,
        SigningAlgorithm="RSASSA_PKCS1_V1_5_SHA_256",
    )
    if not resp.get("SignatureValid"):
        raise AssertionError("Verify: signature not valid")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateKey": CreateKey,
    "DescribeKey": DescribeKey,
    "CreateKmsAlias": CreateAlias,
    "ListKmsAliases": ListAliases,
    "ListKeys": ListKeys,
    "DisableKey": DisableKey,
    "EnableKey": EnableKey,
    "ScheduleKeyDeletion": ScheduleKeyDeletion,
    "CancelKeyDeletion": CancelKeyDeletion,
    "TagKMSResource": TagKMSResource,
    "ListKMSResourceTags": ListKMSResourceTags,
    "UntagKMSResource": UntagKMSResource,
    "Encrypt": Encrypt,
    "Decrypt": Decrypt,
    "GenerateDataKey": GenerateDataKey,
    "GenerateDataKeyWithoutPlaintext": GenerateDataKeyWithoutPlaintext,
    "Sign": Sign,
    "Verify": Verify,
}

SETUP = {
    "kms-keys": setup_kms_keys,
    "kms-crypto": setup_kms_crypto,
    "kms-asymmetric": setup_kms_asymmetric,
}

TEARDOWN = {
    "kms-keys": teardown_kms_keys,
    "kms-crypto": teardown_kms_crypto,
    "kms-asymmetric": teardown_kms_asymmetric,
}
